odl/tomo/data/mrc2014.py

```diff
@@ -0,0 +1,846 @@
+            # section to a list. Return it as a tuple.
+            ext_header = []
+            for section in range(num_sections):
+                entry = {}
+                section_start = section * MRC_FEI_SECTION_SIZE
+                for field in section_fields:
+                    # Get the bytestring from the right position in the file,
+                    # unpack it and store the value as array in the dict.
+                    self.file.seek(section_start + field['offset'])
+                    num_items = int(np.prod(field['dshape']))
+                    size_bytes = num_items * field['dtype'].itemsize
+                    packed_value = self.file.read(size_bytes)
+                    fmt = '{}{}'.format(num_items, field['dtype'].char)
+                    value = struct.unpack(fmt, packed_value)
+                    # Make each entry a 1-element 1D array as usual
+                    entry[field['name']] = np.array(
+                        value, dtype=field['dtype']).reshape(field['dshape'])
+
+                ext_header.append(entry)
+            return tuple(ext_header)
+
+    def read_data(self, dstart=None, dend=None, swap_axes=True):
+        """Read the data from `file` and return it as Numpy array.
+
+        Parameters
+        ----------
+        dstart : int, optional
+            Offset in bytes of the data field. By default, it is equal
+            to ``header_size``. Backwards indexing with negative values
+            is also supported.
+            Use a value larger than the header size to extract a data subset.
+        dend : int, optional
+            End position in bytes until which data is read (exclusive).
+            Backwards indexing with negative values is also supported.
+            Use a value different from the file size to extract a data subset.
+        swap_axes : bool, optional
+            If ``True``, use `data_axis_order` to swap the axes in the
+            returned array. In that case, the shape of the array may no
+            longer agree with `data_shape`.
+
+        Returns
+        -------
+        data : `numpy.ndarray`
+            The data read from `file`.
+        """
+        data = super().read_data(dstart, dend, reshape_order='F')
+        if swap_axes:
+            data = np.transpose(data, axes=self.data_axis_order)
+        return data
+
+    # TODO: read extended header for the standard flavors, see the spec
+    # homepage
+
+
+class FileWriterMRC(FileWriterRawBinaryWithHeader):
+
+    """Writer for the MRC file format.
+
+    See [Che+2015]_ or the `explanations on the CCP4 homepage
+    <http://www.ccpem.ac.uk/mrc_format/mrc2014.php>`_ for the
+    text of the specification.
+
+    References
+    ----------
+    [Che+2015] Cheng, A et al. *MRC2014: Extensions to the MRC format header
+    for electron cryo-microscopy and tomography*. Journal of Structural
+    Biology, 129 (2015), pp 146--150.
+    """
+
+    print_mrc2014_spec = staticmethod(print_mrc2014_spec)
+
```

Comment:
Well, the writer can be handy if you want to store recos in MRC format to process them with some other EM tool. But recos don't need an extended header, and I've never written any tilt series, so I figured the extended header would be close to useless in the writer. If somebody wants to have it, I can add it, no big deal. But not now :-)